### DISCUSSION TITLE
Support for color bitmaps in GraphicDisplay

### DISCUSSION THREAD

**Discussion by mcbridejc** [modm-io/modm#735] [Ideas]
There currently doesn't seem to be a way to draw full color images to a `GraphicDisplay` object. `drawImage` supports monochrome images, drawn with the foreground color, but I wanted to draw a full bitmap with 16-bit color. My current solution has been to simply add a `drawBitmap` method to the class, like so: 

```
/**
 * Draw a bitmap with 16-bit 565 colored pixels.
 *
 * \p data is series of pixel color values
 */
virtual void drawBitmap(glcd::Point upperLeft, uint16_t width, uint16_t height, modm::accessor::Flash<uint8_t> data) = 0;
```

I'm using the ili9341 driver and it already has this method, but of course there are other drivers inheriting from `GraphicDisplay` that do not. I suppose it is technically possible to make a generic implementation which uses `setColor()` and `setPixel()` to write each pixel in the image out, although this is probably very inefficient! 

I'm prepared to do some legwork to get something sorted that works for everyone, but I wanted to first put this out there to see if there are any thoughts on how best to proceed.

**Comment by salkinium** [comment]
The UI code is very old, and while still functioning, is not actively maintained (or used) anymore. As you noted, the access is very inefficient, since it was meant for small 1-bit displays running on AVRs. 

> possible to make a generic implementation which uses setColor() and setPixel() to write each pixel in the image out, although this is probably very inefficient!

You could make that the default implementation in the virtual base class, which would let it be overwritable by the display implementation with something more efficient?

There's been some effort in making a new efficient, scalable graphics system (https://github.com/salkinium/upainter), but I got distracted by more pressing issues. 
There's also been a (failed) PR to integrate LittlevGL into modm: https://github.com/modm-io/modm/pull/362

**Reply by salkinium** [reply]
@TomSaw is working on rewriting this whole part, see https://github.com/modm-io/modm/pull/665

**Comment by TomSaw** [comment]
Hey @mcbridejc,
The whole graphics is going into the next round soon and will be awesome!
There's rich support for colored images, even for indexed colors. I'm currently cutting #665 into smaller pieces and PR them.
May take a few weeks... please be patient, :smiley:

**Reply by mcbridejc** [reply]
Cool. I look forward to having a closer look. I have a [project](https://github.com/mcbridejc/spincoater-controller) that uses a ILI9341 display/touchscreen, and is currently based on a modm fork that I never quite got around to merging back upstream. Which is fine, because it looks like you are doing a much more thorough overhaul. Glad to hear it!
